org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java

```diff
@@ -4571,6 +4574,40 @@ protected void consumeInsideCastExpressionLL1WithBounds() {
 protected void consumeInsideCastExpressionWithQualifiedGenerics() {
 	// InsideCastExpressionWithQualifiedGenerics ::= $empty
 }
+protected void consumeTemplateExpressionWithPrimary() {
+	Expression expression = this.expressionStack[this.expressionPtr--];
+	StringTemplate template;
+	if (expression instanceof StringLiteral) {
+		template = new StringTemplate(new StringLiteral[] {(StringLiteral) expression},
+									new Expression[0], expression.sourceStart, expression.sourceEnd);
+	} else {
+		template = (StringTemplate) expression;
+	}
+	Expression processor = this.expressionStack[this.expressionPtr--];
+	TemplateExpression exp = new TemplateExpression(
+								processor,
+								template);
+	exp.sourceStart = processor.sourceStart;
+	exp.sourceEnd = expression.sourceEnd;
+	pushOnExpressionStack(exp);
+}
+protected void consumeTemplateExpressionWithName() {
+	NameReference ref = getUnspecifiedReference(true);
+	Expression expression = this.expressionStack[this.expressionPtr--];
+	StringTemplate template;
+	if (expression instanceof StringLiteral) {
+		template = new StringTemplate(new StringLiteral[] {(StringLiteral) expression},
+									new Expression[0], expression.sourceStart, expression.sourceEnd);
+	} else {
+		template = (StringTemplate) expression;
+	}
+	TemplateExpression exp = new TemplateExpression(
+								ref,
+								template);
+	exp.sourceStart = ref.sourceStart;
+	exp.sourceEnd = expression.sourceEnd;
+	pushOnExpressionStack(exp);
```

Comment:
Lots of duplication with previous method. Given NameReference is an Expression, perhaps extracting common code may be a possibility